go/vt/servenv/embedder.go

```
@@ -0,0 +1,444 @@
+	mu sync.Mutex
+	f  func(w http.ResponseWriter, r *http.Request)
+}
+
+// Set replaces the existing handler with a new one.
+func (hf *handleFunc) Set(f func(w http.ResponseWriter, r *http.Request)) {
+	hf.mu.Lock()
+	defer hf.mu.Unlock()
+	hf.f = f
+}
+
+// Get returns the current handler.
+func (hf *handleFunc) Get() func(w http.ResponseWriter, r *http.Request) {
+	hf.mu.Lock()
+	defer hf.mu.Unlock()
+	return hf.f
+}
+
+//-----------------------------------------------------------------
+
+// embedder provides the functions needed to embed an object
+// by remapping global endpoints into different namespaces.
+type embedder struct {
+	name, prefix string
+	handleFuncs  map[string]*handleFunc
+	sp           *statusPage
+}
+
+// AssignPrefix assigns a prefix for the instance name. All future
+// stats vars for the instance will be created with the specified
+// prefix. The prefix is is also used to label the additonial
+// dimension for the stats vars.
+func AssignPrefix(instanceName, prefix string) {
+	embedmu.Lock()
+	defer embedmu.Unlock()
+
+	e, ok := embeds[instanceName]
+	if ok {
+		e.resetLocked(prefix)
+		return
+	}
+	e = &embedder{
+		name:        instanceName,
+		prefix:      prefix,
+		handleFuncs: make(map[string]*handleFunc),
+	}
+	if instanceName != "" {
+		e.sp = newStatusPage(instanceName)
+	}
+	embeds[instanceName] = e
+}
+
+func (e *embedder) resetLocked(prefix string) {
+	e.prefix = prefix
+	for _, hf := range e.handleFuncs {
+		hf.Set(nil)
+	}
+
+	for _, vmap := range globalStatVars {
+		vmap.mu.Lock()
+		delete(vmap.vars, e.name)
+		vmap.mu.Unlock()
+	}
+
+	if e.sp != nil {
+		e.sp.reset()
+	}
+}
+
+// findOrCreateEmbedder returns an existing embedder or
+// creates a new one.
+func findOrCreateEmbedder(instanceName string) *embedder {
+	if e, ok := embeds[instanceName]; ok {
+		return e
+	}
+	e := &embedder{
+		name:        instanceName,
+		prefix:      "",
+		handleFuncs: make(map[string]*handleFunc),
+	}
+	if instanceName != "" {
```

I _think_ we should never get into here if instanceName is "", right?

That seems like a coding error and I'd panic instead of creating an empty embedder you never use.